textattack/search_methods/metropolis_hastings.py

```diff
@@ -0,0 +1,175 @@
+import torch
+import math
+import warnings
+import numpy as np
+from textattack.shared import utils
+from textattack.search_methods import SearchMethod
+from transformers import AutoTokenizer, AutoModelWithLMHead
+
+class MetropolisHastingsSampling(SearchMethod):
+    """ 
+    Uses Metropolis-Hastings Sampling to generate adversarial samples.
+    Based off paper "Generating Fluent Adversarial Examples for Natural Langauges" by Zhang, Zhou, Miao, Li (2019)
+
+    Args:
+        max_iter (int): The maximum number of sampling to perform
+        lm_type (str): The language model to use to estimate likelihood of text.
+            Currently supported LM is "gpt-2"
+    """
+
+    def __init__(self, max_iter = 200, lm_type = "gpt-2"):
+        self.max_iter = max_iter
+        self.lm_type = lm_type
+
+        self.lm_tokenizer = AutoTokenizer.from_pretrained("gpt2", use_fast=False)
+        self.language_model = AutoModelWithLMHead.from_pretrained("gpt2")
+
+        try:
+            # Try to use GPU, but sometimes we might have out-of-memory issue
+            # Having the below line prevents CUBLAS error when we try to switch to CPU 
+            torch.cuda.current_blas_handle()
+            self.language_model = self.language_model.to(utils.get_device())
+            self.use_gpu = True
+        except RuntimeError as error:
+            if "CUDA out of memory" in str(error):
+                warnings.warn("CUDA out of memory. Running GPT-2 for Metropolis Hastings on CPU.")
+                self.language_model = self.language_model.to("cpu")
+                self.use_gpu = False
```

Comment:
Okay but `self._lm_device` will have to be manually set to `torch.device('cpu')` if we face CUDA out of memory error because the default `get_device()` will still get the GPU